FirebaseCombineSwift/Sources/Auth/Auth+Combine.swift

```diff
@@ -0,0 +1,485 @@
+    public func updateCurrentUser(_ user: User) -> Future<Void, Error> {
+      Future<Void, Error> { promise in
+        self.updateCurrentUser(user) { error in
+          if let error = error {
+            promise(.failure(error))
+          } else {
+            promise(.success(()))
+          }
+        }
+      }
+    }
+
+    // MARK: - Anonymous Authentication
+
+    /// Asynchronously creates and becomes an anonymous user.
+    ///
+    /// If there is already an anonymous user signed in, that user will be returned instead.
+    /// If there is any other existing user signed in, that user will be signed out.
+    ///
+    /// The publisher will emit events on the **main** thread.
+    ///
+    /// - Returns: A publisher that emits the result of the sign in flow.
+    /// - Remark:
+    ///   Possible error codes:
+    ///   - `AuthErrorCodeOperationNotAllowed` - Indicates that anonymous accounts are
+    ///     not enabled. Enable them in the Auth section of the Firebase console.
+    ///
+    ///   See `AuthErrors` for a list of error codes that are common to all API methods
+    @discardableResult
+    public func signInAnonymously() -> Future<AuthDataResult, Error> {
+      Future<AuthDataResult, Error> { promise in
+        self.signInAnonymously { authDataResult, error in
+          if let error = error {
+            promise(.failure(error))
+          } else if let authDataResult = authDataResult {
+            promise(.success(authDataResult))
+          }
+        }
+      }
+    }
+
+    // MARK: - Email/Password Authentication
+
+    /// Creates and, on success, signs in a user with the given email address and password.
+    ///
+    /// The publisher will emit events on the **main** thread.
+    ///
+    /// - Parameters:
+    ///   - email: The user's email address.
+    ///   - password: The user's desired password.
+    /// - Returns: A publisher that emits the result of the sign in flow.
+    /// - Remark:
+    ///   Possible error codes:
+    ///   - `AuthErrorCodeInvalidEmail` - Indicates the email address is malformed.
+    ///   - `AuthErrorCodeEmailAlreadyInUse` - Indicates the email used to attempt sign up
+    ///     already exists. Call fetchProvidersForEmail to check which sign-in mechanisms the user
+    ///     used, and prompt the user to sign in with one of those.
+    ///   - `AuthErrorCodeOperationNotAllowed` - Indicates that email and password accounts
+    ///     are not enabled. Enable them in the Auth section of the Firebase console.
+    ///   - `AuthErrorCodeWeakPassword` - Indicates an attempt to set a password that is
+    ///     considered too weak. The NSLocalizedFailureReasonErrorKey field in the NSError.userInfo
+    ///     dictionary object will contain more detailed explanation that can be shown to the user.
+    ///
+    ///   See `AuthErrors` for a list of error codes that are common to all API methods
```

Comment:
Agreed. Looking into this a bit, it seems like we'd have to update our error enums in the base SDKs, though. See for example [First-Class Swift API for Objective-C Frameworks | Inside PSPDFKit](https://pspdfkit.com/blog/2018/first-class-swift-api-for-objective-c-frameworks/) and [What’s Right with Errors. You’re enjoying your app when suddenly… | by Bertrand Longevialle | Deezer I/O](https://deezer.io/whats-right-with-errors-2fe419424f62).

Otherwise, we'd have to maintain two (times the number of SDKs) error types - one in the respective base SDK, and one for Firebase Combine.

WDYT?

Reply:
I'm sort of inclined to use this as a testing ground for a new error enum that's specific to Combine, then we could port it back to Auth if it ends up working out well, or revert to the existing error if it doesn't.

Re: maintaining two types, the existing public APIs are very unlikely to get a new error being thrown as it could be a breaking change, it'll have to go through public API proposal so we can catch it then, and overall I think it's a pretty stable API surface.

Reply:
I suggest to go with the current implementation, and implement a Swift-native error type in a second step. Tracked [here](https://github.com/firebase/firebase-ios-sdk/projects/3#card-51035218)